### PR TITLE
PP-790: qmgr output has inconsistent whitespaces

### PR DESCRIPTION
**Pull request by agrawalravi90**
#### Issue-ID

* **[PP-790](https://pbspro.atlassian.net/browse/PP-790)**
* **[EDD](https://pbspro.atlassian.net/wiki/display/PD/PP-790%3A+qmgr+output+has+inconsistent+whitespaces)**
* **[Forum Discussion](http://community.pbspro.org/t/pp-790-qmgr-output-has-inconsistent-whitespaces/566)**

#### Problem description
* qmgr output has inconsistent whitespaces

#### Cause / Analysis
* For all commands (qstat, pbsnodes etc.), the code uses spaces (" ") as prefix for new attributes, but for qmgr it uses tabs. Similarly, for line extensions in qmgr, it uses spaces as prefix while the other commands use tabs. This ticket was basically intended to make qmgr's output consistent with other commands.

#### Solution description
* Modified the code inside display() to add 4 spaces as prefix for new attributes and tabs as prefix for line extensions, which is what qstat and pbsnodes use.
* Also added a new test suite TestQmgr and added a PTL test to check for whitespace consistency.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by prakashcv13** [comment]
Ravi,

Wont we need external interface change document for this?

Thanks,
Prakash

**Comment by agrawalravi90** [comment]
I was in two minds about it, but if you think that this needs an interface change document as well then I'll create one. Thanks.

**Comment by agrawalravi90** [comment]
Ok, I just started a forum discussion and created the EDD, please see description of the PR for links. Thanks.

**Comment by agrawalravi90** [comment]
Requesting reviews, this is a pretty small change, thanks a bunch!

File: test/tests/functional/pbs_qmgr.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software - under
+# a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+import os
+from string import ascii_letters
```
Comment (arungrover) [inline test/tests/functional/pbs_qmgr.py]: I do not see these two modules being used in this ptl script

Comment (agrawalravi90) [inline test/tests/functional/pbs_qmgr.py]: I use os to remove the temporary file created, but you are right about ascii_letters, I'll remove it.

File: test/tests/functional/pbs_qmgr.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            # Tabs are prefix for line extensions
+            # Make sure that this is a line extension
+            self.assertTrue("=" not in line)
+
+    def test_listcmd_whitespaces(self):
+        """
+        Check that the prefix for new attributes listed out by qmgr list
+        are spaces and that for line extensions is a tab
+        """
+        pass
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_qmgr.py]: oh wow, that's a bug, that explains why the test would pass with 'check_whitespace_prefix' instead of '__check_whitespace_prefix', thanks for catching this.

File: test/tests/functional/pbs_qmgr.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        # Execute "qmgr 'list node vn[0]'"
+        # The comment attribute should generate a line extension
+        if self.du.is_localhost(self.server.hostname) is True:
+            qmgr_cmd = ["qmgr", "-c", "list node " + nodename]
+        else:
+            qmgr_cmd = ["qmgr", "-c", "\'list node " + nodename + "\'"]
+        with open(fn, "w+") as tempfd:
+            ret = self.du.run_cmd(self.server.hostname, qmgr_cmd,
+                                  stdout=tempfd)
+            self.assertTrue(ret['rc'] == 0)
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_qmgr.py]: hmm, thanks for pointing that out, it will not get handled. I'll wrap the code inside a try-finally block to make sure that cleanup happens regardless of exceptions.

File: test/tests/functional/pbs_qmgr.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            for line in tempfd:
+                self.__check_whitespace_prefix(line)
+
+        # Cleanup
+        # Remove the temporary file
+        os.remove(fn)
+        # Delete the vnode created
+        self.mom.delete_vnode_defs(vndef_file)
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_qmgr.py]: Actually this would leave the vnode as it as when I'd run this test. Maybe it's because this suite has only one test case so setup() doesn't get called again to clean stuff up.

**Comment by agrawalravi90** [comment]
I've addressed your comments Arun, please provide further feedback, thanks!

File: test/tests/functional/pbs_qmgr.py [hunk 5]
```diff
@@ -37,10 +37,10 @@
 # trademark licensing policies.
 
 import os
-from string import ascii_letters
 from random import choice
```
Comment (arungrover) [inline test/tests/functional/pbs_qmgr.py]: In my previous comment I meant to point 'from random import choice' as the second redundant import :)

File: test/tests/functional/pbs_qmgr.py [hunk 6]
```diff
@@ -72,57 +72,56 @@ def test_listcmd_whitespaces(self):
...
+            # Set 'comment' attribute to the long string we created above
+            attrs = {ATTR_comment: long_string}
+            self.server.manager(MGR_CMD_SET, VNODE, attrs, nodename)
+            # Execute "qmgr 'list node vn[0]'"
+            # The comment attribute should generate a line extension
+            if self.du.is_localhost(self.server.hostname) is True:
+                qmgr_cmd = ["qmgr", "-c", "list node " + nodename]
+            else:
+                qmgr_cmd = ["qmgr", "-c", "\'list node " + nodename + "\'"]
+            tempfd = open(fn, "w+")
```
Comment (arungrover) [inline test/tests/functional/pbs_qmgr.py]: just curious why you are opening file in w+ mode here but in write mode in the previous open call.

Comment (agrawalravi90) [inline test/tests/functional/pbs_qmgr.py]: They should both be "w+" as we write and read both, thanks for catching this.

**Comment by agrawalravi90** [comment]
@arungrover I addressed your comment, also fixed some bugs that I had introduced with my previous commit (sorry that i forgot to test them), please let me know if things look okay now. Thanks!

File: test/tests/functional/pbs_qmgr.py [hunk 7]
```diff
@@ -98,11 +100,19 @@ def test_listcmd_whitespaces(self):
             long_string += blah
             # Create a new vnode
             attrs = {ATTR_rescavail: 2}
```
Comment (arungrover) [inline test/tests/functional/pbs_qmgr.py]: I missed to add a comment here in last round. ATTR_rescavail is just "resoures_available" you do need ".<resource_name>" appended to it to make it meaningful. I am a little surprised to know that pbs_mom is conveniently ignoring such abnormality in vnodedef file.

File: test/tests/functional/pbs_qmgr.py [hunk 8]
```diff
@@ -98,11 +100,19 @@ def test_listcmd_whitespaces(self):
...
-            vndef = self.mom.create_vnode_def("vn", attrs)
+            vndef = self.mom.create_vnode_def(node_prefix, attrs)
```
Comment (arungrover) [inline test/tests/functional/pbs_qmgr.py]: If the purpose here is just to add one vnode then please use self.server.create_vnodes(nodename, attrs, 1, self.mom). This will make sure that vnode is created with attributes you specified.
If you do this then you don't have to check for mom to be up

**Comment by agrawalravi90** [comment]
Thanks Arun, I addressed your comments but, I apologize, I accidentally did a git --amend on the previous commit, so the new changes got merged with the old ones, I hope it won't be too much of a pain to review them. Please let me know if things look okay now.

File: test/tests/functional/pbs_qmgr.py [hunk 9]
```diff
@@ -37,7 +37,7 @@
 # trademark licensing policies.
 
 import os
-from random import choice
+import time
 
 from tests.functional import *
 from ptl.lib.pbs_ifl_mock import *
```
Comment (arungrover) [inline test/tests/functional/pbs_qmgr.py]: This import is not needed. it should get imported with tests.functional.

**Comment by agrawalravi90** [comment]
Thanks Arun, I've removed the import and squashed the commits, requesting @PBSPro/pbspro-maintainers to merge this in if all looks good. Thanks!

**Comment by arungrover** [comment]
Thanks for making changes @agrawalravi90. Please post the result of latest PTL script on the ticket. Let's wait for one more review and then it's ready for merge.

**Comment by agrawalravi90** [comment]
@arungrover sorry for the delay, I've attached the latest logs.

@PBSPro/pbspro-maintainers Arun has already signed-off on this, can one of you please review this and merge it in? It's a pretty small change. Thanks!

File: src/cmds/qmgr.c [hunk 10]
```diff
@@ -2211,25 +2211,29 @@ display(int otype, int ptype, char *oname, struct batch_status *status,
...
 			}
 			else  {
+				indent_len = 4;
 				if (otype == MGR_OBJ_RSC) {
 					if ((attr != NULL) && (strcmp(attr->name, "type") == 0)) {
 						struct resc_type_map *rtm = find_resc_type_map_by_typev(atoi(attr->value));
 						if (rtm) {
-							printf("\ttype = %s\n", rtm->rtm_rname);
+							printf("%*s", indent_len, " ");
+							printf("type = %s\n", rtm->rtm_rname);
```
Comment (bhroam) [inline src/cmds/qmgr.c]: Did you have a reason to make these two printf()'s instead of one?  It's just one line of output

File: test/tests/functional/pbs_qmgr.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        Check whether the whitespace prefix for the line specified is correct
+
+        :param line: the line to check
+        :type line: String
+        """
+        if line is None:
+            return
+
+        if line[0] == " ":
+            # Spaces are prefix for new attribute lines
```
Comment (bhroam) [inline test/tests/functional/pbs_qmgr.py]: I think you need the word 'the' in front of prefix.

File: test/tests/functional/pbs_qmgr.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+            if self.du.is_localhost(self.server.hostname) is True:
+                qmgr_cmd = ["qmgr", "-c", "list sched"]
+            else:
+                qmgr_cmd = ["qmgr", "-c", "\'list sched\'"]
+            with os.fdopen(fd, "w+") as tempfd:
+                ret = self.du.run_cmd(self.server.hostname, qmgr_cmd,
+                                      stdout=tempfd)
+
+                self.assertTrue(ret['rc'] == 0)
+                for line in tempfd:
```
Comment (bhroam) [inline test/tests/functional/pbs_qmgr.py]: Consider using ret['out'] instead of writing the output to a file and reading the file back in.

File: test/tests/functional/pbs_qmgr.py [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+                    self.__check_whitespace_prefix(line)
+
+        finally:
+            # Cleanup
+            # Remove the temporary file
+            os.remove(fn)
+            # Delete the vnode created
+            if vndef_file is not None:
+                self.mom.delete_vnodes()
+                self.server.manager(MGR_CMD_DELETE, VNODE, id=nodename)
```
Comment (bhroam) [inline test/tests/functional/pbs_qmgr.py]: This will happen in revert_to_defaults().  You don't need to do it.

File: src/cmds/qmgr.c (hunk 10 above)
Comment (agrawalravi90) [inline src/cmds/qmgr.c]: it was getting a bit confusing because of the "%*s" as that single formatter requires two arguments (the length and the blank space). So, i thought it'd be more clear if I separated it out. But if you think it doesn't make much difference, I'll combine them into one.

File: test/tests/functional/pbs_qmgr.py (hunk 12 above)
Comment (agrawalravi90) [inline test/tests/functional/pbs_qmgr.py]: ret['out'] doesn't return the 'pristine' output of the command, it returns a list of the contents, and because I'm looking for something as subtle as whitespaces in the output, I wanted to look at the pristine, un-process output. That's why I think it's better to just write it out to a file and read it back in. What do you think?

File: test/tests/functional/pbs_qmgr.py (hunk 13 above)
Comment (agrawalravi90) [inline test/tests/functional/pbs_qmgr.py]: True, but revert_to_defaults() is executed only during setUp(), this means that if one just ran a single test case, whatever setUp() it had created will linger on until you run another test case.

**Comment by agrawalravi90** [comment]
Thanks for the review Bhroam! I've addressed your comments, please provide further feedback.

File: src/cmds/qmgr.c (hunk 10 above)
Comment (bhroam) [inline src/cmds/qmgr.c]: I wouldn't find it confusing, but if you do, let's err on the side of readability and leave it as twl.

File: test/tests/functional/pbs_qmgr.py (hunk 12 above)
Comment (bhroam) [inline test/tests/functional/pbs_qmgr.py]: True, whitespace is rather subtle.  It's easy for something to process it away without thinking much of it.  Writing out to a file is probably best.

File: test/tests/functional/pbs_qmgr.py (hunk 13 above)
Comment (bhroam) [inline test/tests/functional/pbs_qmgr.py]: setUp() is called before every test.  No matter how many tests you run, revert_to_defaults() will be run.  The only thing this will do is delete it prior to leaving the final test.  I'd rather you leave it to revert_to_defaults().  It's the standard place this sort of thing is done in PTL.

**Comment by bhroam** [comment]
I sign off.  Go ahead and squash and rebase for the final merge.

**Comment by agrawalravi90** [comment]
Thanks bhroam, I've squashed the commits.
